Third-party Cart integration

cbj4074 opened the pull request:
This PR aims to integrate the https://github.com/bumbummen99/LaravelShoppingcart package.

Note: I based this on top of another branch because it requires the changes therefrom, but I'll rebase before we merge.

cbj4074 committed:
feat: Integrate third-party cart package

poppabear8883 commented on database/factories/ShoppingCartFactory.php:
```diff
@@ -0,0 +1,20 @@
+use IndieHD\Velkart\Contracts\ShoppingCartRepositoryContract;
+use Ramsey\Uuid\Uuid;
+
+$cart = resolve(Cart::class);
```
Im either blind or you are not using `$cart` anywhere ?

cbj4074 replied:
You are correct; I changed the way this works and neglected to remove that. Thanks!

poppabear8883 commented on database/migrations/2020_01_12_000000_create_shoppingcart_table.php:
```diff
@@ -11,20 +11,21 @@ class CreateShoppingcartTable extends Migration
-        Schema::create('shoppingcart', function (Blueprint $table) {
+        Schema::create('shopping_carts', function (Blueprint $table) {
+            $table->bigIncrements('id');
             $table->string('identifier');
```
should this be indexed ?

cbj4074 replied:
Yes, it probably should. Even though we're not joining on it (we're joining on the `id`), we still need to be able to locate a cart instance by this column. Good call.

cbj4074 replied:
Oh, actually, the `identifier` column is part of the compound unique index, too, so a separate index is redundant. No change required, then.

poppabear8883 commented on database/migrations/2020_01_12_000000_create_shoppingcart_table.php:
```diff
@@ -11,20 +11,21 @@ class CreateShoppingcartTable extends Migration
+        Schema::create('shopping_carts', function (Blueprint $table) {
+            $table->bigIncrements('id');
             $table->string('identifier');
             $table->string('instance');
```
This might also need to be indexed ?

cbj4074 replied:
It does, but the subsequent `$table->unique(['identifier', 'instance']);` covers it.

poppabear8883 commented on database/factories/OrderFactory.php:
```diff
@@ -1,16 +1,15 @@
+use IndieHD\Velkart\Contracts\ShoppingCartRepositoryContract;
 
 $order = resolve(OrderRepositoryContract::class);
+$cart = resolve(ShoppingCartRepositoryContract::class);
```
Do we really want our cart implementation be named `ShoppingCart` vs simply just `Cart` ? It will be in a different namespace so shouldn't be concerned with conflicts.

cbj4074 replied:
That's a fair point. Shorter is better. I will change it.

poppabear8883 commented on tests/Feature/Repositories/ShoppingCartTest.php:
```diff
@@ -0,0 +1,54 @@
+
+use IndieHD\Velkart\Contracts\OrderRepositoryContract;
+use IndieHD\Velkart\Contracts\ShoppingCartRepositoryContract;
+use Ramsey\Uuid\Uuid;
```
Is this dependency inject-able using a contract by chance ? This is definitely an implementation that has better odds of being swapped out later so it may make sense to use it in such a way.

cbj4074 replied:
Good question. I had the same thought initially and I did look into it but it wasn't straightforward because there seems to be a constructor that requires arguments:

```
Illuminate\Contracts\Container\BindingResolutionException : Unresolvable dependency resolving [Parameter #0 [ <required> array $fields ]] in class Ramsey\Uuid\Uuid
```

I'll dig a little deeper.

cbj4074 replied:
I just pushed a correction for this. Thanks again.

mblarsen replied:
FYI it is available through `Illuminate\Support\Str` — e.g. `Str::uuid()` but I understand if you'd want it injected.

poppabear8883 replied:
@mblarsen

I think he's using the 3rd party UUID package for other reasons then just generating unique id's ? Forgive me if I sound ignorant. Its been a minute ;)

cbj4074 replied:
That's correct @poppabear8883 . It's been a while since I've looked at this, but my recollection is that I used the `Ramsey` library because it provides more control over the UUID generation, and more specifically, provides for the various UUID versions (particularly version 4, which is pseudo-random).

poppabear8883 reviewed (commented):
A couple comments, i'll review again after those have been resolved.

cbj4074 committed:
refact: Change ShoppingCart (and all references) to just Cart

Shorter is better and there's no compelling reason to include the
"Shopping" everywhere.

cbj4074 committed:
refact: Eliminate static calls to uuid4() method

poppabear8883 commented on database/factories/CartFactory.php:
```diff
@@ -2,13 +2,13 @@
 $factory->define($cartRepository->modelClass(), function (Faker\Generator $faker) {
 
-    $identifier = Uuid::uuid4();
+    $identifier = resolve(UuidFactoryInterface::class)->uuid4();
```
if you resolve here it will resolve on every iteration, i would recommend taking this out of the callback

poppabear8883 replied:
I take that back, i suppose it will regardless ?

cbj4074 replied:
You're correct; that should be moved outside of the closure, regardless of the factory's inner-workings. There is no good reason to resolve it inside, and it probably hurts performance one way or another.

cbj4074 replied:
I just pushed a fix.

cbj4074 committed:
fix: Remove redundant index

This field is part of the unique compound index defined subsequently.

cbj4074 committed:
refact: Move interface resolution outside of closure

There is no good reason to resolve inside the closure, and doing so
probably hurts performance in some capacity.

cbj4074 commented on database/migrations/2020_01_12_000000_create_cart_table.php:
```diff
@@ -1,30 +1,31 @@
             $table->nullableTimestamps();
 
-            $table->primary(['identifier', 'instance']);
+            $table->unique(['identifier', 'instance']);
```
Hmm, I'm not entirely sure why I changed this from `primary()` to `unique()`. I don't think I would have done it arbitrarily, though.

cbj4074 replied:
Ahh, yes. I added `$table->bigIncrements('id')` to this table, in which case `primary(...)` had to be changed to `unique(...)`, because the table can only have one primary key (obviously).

cbj4074 commented on tests/Feature/Repositories/CartTest.php:
```diff
@@ -0,0 +1,61 @@
+    {
+        $order = factory($this->order->modelClass())->create();
+
+        $cart = $this->getRepository()->findById($order->cart->id);
```
This approach is not intuitive. I was being lazy and didn't rig-up the `Cart` factory to have a `withOrder` state, so this test is relying on the `Order` factory (which creates both related models). So, while it's "accurate", it's confusing. I should rework this.

cbj4074 committed:
refact: Employ more straightforward logic

poppabear8883 reviewed (approved):
lgtm
